### PR TITLE
Add utilities for floris models

### PR DESCRIPTION
**Pull request by paulf81**
# Add utilities for floris models

co-authored with @bayc 

This pull request:

- Adds helper functions to utilities to work with nested dictionaries
- Add methods to the FlorisModel for getting/setting parameters using those functions
- Adds methods for changing the power-thrust model
- Adds testing for all new functions
- Reorganizes FlorisModel for clarity

### CODE REVIEW

File: floris/floris_model_utils.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+import numpy as np
+import yaml
+from attrs import define, field
+
+from floris import FlorisModel
+from floris.type_dec import floris_array_converter, NDArrayFloat
+
+
+def nested_get(dic: Dict[str, Any],
```
Comment (rafmudaf) [inline floris/floris_model_utils.py:22]: Throughout FLORIS, the style convention for split line arguments is to open the parenthesis, then list the arguments each on their own line, and then close the parenthesis on the final line:
```python
def nested_get(
    d: Dict[str, Any],
    keys: List[str]
) -> Any:
```

File: floris/floris_model_utils.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+               keys: List[str]) -> Any:
+    """Get a value from a nested dictionary using a list of keys.
+    Based on: stackoverflow.com/questions/14692690/access-nested-dictionary-items-via-a-list-of-keys
+
+    Args:
+        dic (Dict[str, Any]): The dictionary to get the value from.
```
[the remark this answers is not in the document]
Comment (paulf81) [reply floris/floris_model_utils.py:28]: hahahahaha

File: floris/floris_model_utils.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        value (Any): The value to set.
+        idx (Optional[int], optional): If the value is an list, the index to change.
+         Defaults to None.
+    """
+    dic_in = dic.copy()
+
+    for key in keys[:-1]:
+        dic = dic.setdefault(key, {})
+    if idx is None:
+        # Parameter is a scaler, set directly
```
[the remark this answers is not in the document]
Comment (paulf81) [reply floris/floris_model_utils.py:59]: fixed

File: floris/floris_model_utils.py (hunk 1 above)
Comment (paulf81) [reply floris/floris_model_utils.py:22]: fixed

File: floris/floris_model_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulf81) [reply floris/floris_model_utils.py:28]: Fixed

**Comment by paulf81** [comment]
> Architecturally, @paulf81 what do you think of putting `nested_get`, `nested_set`, and `print_nested_dict` in `floris.utilities` since they are general function, and the other functions that apply only to `FlorisModel` within the floris_model module?

Done!

**Comment by paulf81** [comment]
> @paulf81 , agree with Raf's comments above. Also for reference, in this commit ([cbd3cf8](https://github.com/NREL/floris/commit/cbd3cf87f079bb800a44a577a5943bce822c28cd)), I removed a slightly different version of this same functionality where the list of keys was optional, and if not provided, the object would be scraped for all attributes and those would be returned as a dictionary. Could be worth including if you think it is useful.

Sounds interesting!  Looking at the code I wasn't quite positive which function should be added back in, and then I think some of these will need an update to the v4 standards, do you want to try to add into this PR?  Or a seperate one?

**Comment by bayc** [comment]
> > Architecturally, @paulf81 what do you think of putting `nested_get`, `nested_set`, and `print_nested_dict` in `floris.utilities` since they are general function, and the other functions that apply only to `FlorisModel` within the floris_model module?
> 
> Done!

Hey @paulf81 , I think @rafmudaf was suggesting moving the `get_fmodel_param`, `set_fmodel_param`, and `print_fmodel_dict` functions onto the `FlorisModel` class itself. I have opened up a PR to your branch that does this here: https://github.com/paulf81/floris/pull/13. If you both can check it out and see if 1) this is what you meant @rafmudaf , and 2) @paulf81 you agree with the changes, then feel free to merge it @paulf81 .

**Comment by paulf81** [comment]
@misi9170 , I added some code for getting/setting the power_thurst_model, and somehow it doesn't work.  It may be this thing of how the turbine_type can be stored in a few possible formats but I was wondering if you might have a look at these new codes Monday and see if you spot my error?  I added two simple tests: and this one is failing:

```
def test_set_power_thrust_model():

    fmodel = FlorisModel(configuration=YAML_INPUT)
    fmodel = set_power_thrust_model(fmodel, "simple-derating")
    assert get_power_thrust_model(fmodel) == "simple-derating"

```

**Comment by bayc** [comment]
> @misi9170 , I added some code for getting/setting the power_thurst_model, and somehow it doesn't work. It may be this thing of how the turbine_type can be stored in a few possible formats but I was wondering if you might have a look at these new codes Monday and see if you spot my error? I added two simple tests: and this one is failing:
> 
> ```
> def test_set_power_thrust_model():
> 
>     fmodel = FlorisModel(configuration=YAML_INPUT)
>     fmodel = set_power_thrust_model(fmodel, "simple-derating")
>     assert get_power_thrust_model(fmodel) == "simple-derating"
> ```

@paulf81 , @misi9170  I went ahead and added this functionality and test to my [PR](https://github.com/paulf81/floris/pull/13) to your branch Paul in this commit: https://github.com/paulf81/floris/pull/13/commits/afad340f94edcc176756385b48ec53ee3a5a3678. I changed it so that is uses the currently loaded turbine type, updates the `power_thrust_model`, and passes that updated turbine type to the `set` method. This way we avoid an file I/O and this would retain any other changes made programmatically by the use to the turbine definition.

**Comment by paulf81** [comment]
> > @misi9170 , I added some code for getting/setting the power_thurst_model, and somehow it doesn't work. It may be this thing of how the turbine_type can be stored in a few possible formats but I was wondering if you might have a look at these new codes Monday and see if you spot my error? I added two simple tests: and this one is failing:
> > ```
> > def test_set_power_thrust_model():
> > 
> >     fmodel = FlorisModel(configuration=YAML_INPUT)
> >     fmodel = set_power_thrust_model(fmodel, "simple-derating")
> >     assert get_power_thrust_model(fmodel) == "simple-derating"
> > ```
> 
> @paulf81 , @misi9170 I went ahead and added this functionality and test to my [PR](https://github.com/paulf81/floris/pull/13) to your branch Paul in this commit: [paulf81@afad340](https://github.com/paulf81/floris/commit/afad340f94edcc176756385b48ec53ee3a5a3678). I changed it so that is uses the currently loaded turbine type, updates the `power_thrust_model`, and passes that updated turbine type to the `set` method. This way we avoid an file I/O and this would retain any other changes made programmatically by the use to the turbine definition.

Thank you @bayc , I merged your pr into mine so those changes are now reflected in this pull request

**Comment by paulf81** [comment]
@rafmudaf , I merged in @bayc changes and I think your requested changes are in

@misi9170 , have a look at the new change power-thrust model @bayc implemented, it's a lot shorter than previous and just wanted to make sure I'm not forgetting a reason we went back to the full yaml file before

**Comment by misi9170** [comment]
@paulf81 @bayc 

Those look fine to me. For some reason, @paulf81 , I can't see the history of what you had implemented earlier that wasn't working? But perhaps the fix was to add the `self.set(turbine_type=[turbine_type])` line---this is definitely required for the `FlorisModel` to be reinstantiated, which will be needed to reassign the operation model.

**Comment by paulf81** [comment]
> @paulf81 @bayc
> 
> Those look fine to me. For some reason, @paulf81 , I can't see the history of what you had implemented earlier that wasn't working? But perhaps the fix was to add the `self.set(turbine_type=[turbine_type])` line---this is definitely required for the `FlorisModel` to be reinstantiated, which will be needed to reassign the operation model.

I think that's now moot, it was the change power-thrust functions I had written, but Chris' versions don't seem to suffer from this problem

**Comment by misi9170** [comment]
Overall this looks fine to me, but so that we have a record and some guidelines for future development, what is the logic behind the reorganization of the `FlorisModel` class?

The only objection I have is that I would have kept `reset_operation()` closer to `set()` and `_reinitialize()`, but perhaps there's a good reason to have that moved down!

**Comment by paulf81** [comment]
> Overall this looks fine to me, but so that we have a record and some guidelines for future development, what is the logic behind the reorganization of the `FlorisModel` class?
> 
> The only objection I have is that I would have kept `reset_operation()` closer to `set()` and `_reinitialize()`, but perhaps there's a good reason to have that moved down!

@bayc would you mind to comment here?

**Comment by rafmudaf** [comment]
> The only objection I have is that I would have kept reset_operation() closer to set() and _reinitialize(), but perhaps there's a good reason to have that moved down!

It's subtle, but I also suggest to put `set()`, `reset_operation()`, and `_reinitialize()` above `run()` since they'll always be used in that order.

**Comment by bayc** [comment]
@misi9170 @rafmudaf , after dividing into subsections, I just went with alphabetical order, starting with public methods and ending with private methods. Happy to rearrange as you see fit though.

**Comment by misi9170** [comment]
I definitely like the idea of having the different methods grouped. I've reorganized them again slightly and pushed up the change (which we can certainly revert), so that now we have:

0. `__init__()`
### Methods for setting and running the Floris Model
1. `_reinitialize()`
2. `_set_operation()`
3. `set()`
4. `reset_operation()`
5. `run()`
6. `run_no_wake()`

###

**Comment by bayc** [comment]
That organization works for me @misi9170 !

**Pull request merged**